File: nn/network.m
# Neural network to recognize hand written digits

fprintf('Running digit recognition for neural network.\n');

## Setup the parameters you will use for this exercise
input_layer_size  = 784;  % 28x28 Input Images of Digits
hidden_layer_size = 25;   % 25 hidden units
num_labels = 10;          % 10 labels, from 1 to 10   
lambda = 0;

## Load data
fprintf('Loading and Visualizing Data ...\n')
data = csvread('../data/train.csv');
y = data(:,1);
X = data(:,2:end);
X = [ones(size(X,1),1),X]; # Add ones column 
clear data;     % Remove data variable

## Randomly initialize theta
Theta1 = stdnormal_rnd(hidden_layer_size, input_layer_size + 1);
Theta2 = stdnormal_rnd(num_labels, hidden_layer_size + 1);

## Implement sigmoid

## Implement feedfwd
a2 = feedfwd(X, Theta1);
a2 = [ones(size(a2,1),1),a2]; # Add ones column
hx = feedfwd(a2, Theta2);

## Train the theta

initial_nn_params = [Theta1(:);Theta2(:)]; # unroll the theta

### maximum no of iterations before optimization stops.
options = optimset('MaxIter',50);

costFunction = @(p) nnCostFunction(X, Theta1, Theta2, y,num_labels, ...
				   lambda);
### find local min
[nn_params, cost] = fminunc(costFunction, initial_nn_params, options);

### reshape Theta from nn_params
Theta1 = reshape(nn_params(1:hidden_layer_size * (input_layer_size + ...
						  1)), ...
		 hidden_layer_size, (input_layer_size + 1));

Theta2 = reshape(nn_params((1 + (hidden_layer_size * (input_layer_size ...
						      + 1))):end), ...
		 num_labels, (hidden_layer_size + 1)); 
### Predict
fprintf('Predicting.\n');
pred = predict(Theta1, Theta2, X(:,2:end));

fprintf('\nTraining Set Accuracy: %f\n', mean(double(pred == y)) * 100);

File: R/network.R
# Digit Recognition code using neural network
setwd('/Users/snowleopard/Documents/workspace/kaggle/digit-recognition/R/');
print('Running digit recognition for neural network')

closeAllConnections()
#rm(list=ls())

## import functions
source('sigmoid.R')
#source('costFunction.R')
source('cost.R')
source('gradient.R')
source('predict.R')
## initialize paramters
input_layer_size = 784; # 28x28 input images
hidden_layer_size = 25
num_labels = 10
lambda = 1

## Load datadi
print("Loading data")
# data = read.csv('../data/train.csv')
# y = data[,1]
# x = data[,2:dim(data)[2]]
# x = as.matrix(cbind(1, x)) # add ones column
# rm(data) # remove original data object from memory
m = dim(x)[1]

## Randomly initialize theta between 0 and 1
print('Initializing theta...')
Theta1 = replicate(input_layer_size + 1, rnorm(hidden_layer_size, mean = 0, sd = 1))
Theta2 = replicate(hidden_layer_size + 1, rnorm(num_labels, mean = 0, sd = 1))

# Theta1 = matrix(data=0, ncol=input_layer_size + 1, nrow=hidden_layer_size)
# Theta2 = matrix(data=0, ncol=hidden_layer_size + 1, nrow=num_labels)

## Feed forward algo
print('Runnning feedfwd...')
a2 = sigmoid(x %*% t(Theta1))
a2 = cbind(1, a2) # add ones col
hx = sigmoid(a2 %*% t(Theta2))

## convert y lables to binary vector
yvec = 1 * outer(y, 0:(num_labels-1), FUN = "==")

## Train the theta
print('Running costFunction...')
initial_theta = t(c(Theta1, Theta2))

#costFunction()
#  res <- optim(par = initial_theta, fn = cost, gr = gradient, method = "CG", 
#               control = list(maxit=50))
#library(optimx)
# res <- optim(par = initial_theta, fn = cost,control = list(trace=1),
#               method = c("CG"),hidden_layer_size=hidden_layer_size, 
#               input_layer_size=input_layer_size, 
#               num_labels=num_labels, hx=hx, yvec=yvec)



## - --------------
for(i in 1:4) {
  theta = gradient(initial_theta, hidden_layer_size, input_layer_size, num_labels, hx, yvec)
  J = cost(theta, hidden_layer_size, input_layer_size, num_labels, hx, yvec);
  
}
## reshape theta from par
Theta1 = matrix(theta, nrow = hidden_layer_size, ncol = (input_layer_size + 1))
Theta2 = matrix(theta, nrow = num_labels, ncol = (hidden_layer_size + 1))

## ----------------

p1 = predict()
print(paste('accuracy ', mean((p1 == y)) * 100))

